samples/constant_pH_ensemble_complex_reaction.py

```diff
@@ -0,0 +1,168 @@
+#
+# Copyright (C) 2013-2018 The ESPResSo project
+#
+# This file is part of ESPResSo.
+#
+# ESPResSo is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# ESPResSo is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+"""
+Guide for the constant pH ensemble. The modeled reaction is
+:math:`2\\mathrm{A} + 3\\mathrm{B} \\leftrightarrow 4\\mathrm{C} + 1\\mathrm{H} + 3\\mathrm{E}`, where H represents a proton. In th constant pH ensemble an implicit coupling to a proton reservoir is assumed.
+"""
+
+import pprint
+import numpy as np
+import scipy.optimize
+
+import espressomd
+from espressomd import reaction_ensemble
+
+# System parameters
+#############################################################
+box_l = 50.0
+volume = box_l**3
+
+# Integration parameters
+#############################################################
+system = espressomd.System(box_l=[box_l] * 3)
+np.random.seed(seed=343)
+
+system.time_step = 0.02
+system.cell_system.skin = 0.4
+
+# Particle setup
+#############################################################
+# 2A +3B <-> 4C + 1H +3 E
+N0 = 50  # number of reactant pairs
+
+type_A = 0
+type_B = 1
+type_C = 2
+type_H = 3
+type_E = 4
+types = [type_A, type_B, type_C, type_H, type_E]
+types_name = {type_A: 'A', type_B: 'B', type_C: 'C', type_H: 'H', type_E: 'E'}
+
+nu_A = -2
+nu_B = -3
+nu_C = 4
+nu_H = 1
+nu_E = 3
+
+nu_s = [nu_A, nu_B, nu_C, nu_H, nu_E]
+nu_bar = np.sum(nu_s)
+
+# reaction constant
+K = 10**-3
+# reference concentration for which the reaction constant is reported, 1 mol/l
+c_ref_in_mol_per_l = 1.0
+# simulation units: 1 sigma = 3.55 Angstrom
+conversion_inv_sigma_cube_to_mol_per_l = 37.1
+c_ref_in_1_div_sigma_cubed = c_ref_in_mol_per_l / \
+    conversion_inv_sigma_cube_to_mol_per_l
+pH=10.0
+
+# setup N0 pairs of reactants
+for i in range(N0):
+    for j in range(abs(nu_A)):
+        system.part.add(pos=np.random.random(3) * system.box_l, type=type_A)
+    for j in range(abs(nu_B)):
+        system.part.add(pos=np.random.random(3) * system.box_l, type=type_B)
+
+# use an exclusion radius of 0 to simulate an ideal gas
+RE = reaction_ensemble.ConstantpHEnsemble(
+    temperature=1, exclusion_radius=0, seed=4, conversion_factor_from_mol_per_l_to_1_div_sigma_cubed=1.0/37.1)
+
+RE.add_reaction(
+    gamma=K*(c_ref_in_1_div_sigma_cubed)**nu_bar, 
```

Comment:
Since conversion_factor_from_mol_per_l_to_1_div_sigma_cubed is now passed to the reaction method, one could do this conversion internally and only provide the dimensionless equilibrium constant K = prod a_i**nu_i to the reaction method. This would probably simplify the useage of the reaction methods.

Reply:
I implemented the changes, gamma which is passed to the reaction methods is now the dimenstionless reaction constant.
The discussion about gamma in https://espressomd.github.io/doc/advanced_methods.html#reaction-ensemble needs to be adapted.